src/core_functions/scalar/map/map_entries.cpp

```diff
@@ -36,25 +55,24 @@ static unique_ptr<FunctionData> MapEntriesBind(ClientContext &context, ScalarFun
 		return nullptr;
 	}
 
+	if (map.id() == LogicalTypeId::SQLNULL) {
+		// Input is NULL, output is STRUCT(NULL, NULL)[]
+		auto map_type = LogicalType::MAP(LogicalTypeId::SQLNULL, LogicalTypeId::SQLNULL);
+		bound_function.return_type = CreateReturnType(map_type);
+		return make_uniq<VariableReturnBindData>(bound_function.return_type);
+	}
+
 	if (map.id() != LogicalTypeId::MAP) {
 		throw InvalidInputException("The provided argument is not a map");
 	}
-	auto &key_type = MapType::KeyType(map);
-	auto &value_type = MapType::ValueType(map);
-
-	child_types.push_back(make_pair("key", key_type));
-	child_types.push_back(make_pair("value", value_type));
-
-	auto row_type = LogicalType::STRUCT(child_types);
-
-	bound_function.return_type = LogicalType::LIST(row_type);
+	bound_function.return_type = CreateReturnType(map);
 	return make_uniq<VariableReturnBindData>(bound_function.return_type);
 }
 
 ScalarFunction MapEntriesFun::GetFunction() {
 	//! the arguments and return types are actually set in the binder function
 	ScalarFunction fun({}, LogicalTypeId::LIST, MapEntriesFunction, MapEntriesBind);
-	fun.null_handling = FunctionNullHandling::DEFAULT_NULL_HANDLING;
+	fun.null_handling = FunctionNullHandling::SPECIAL_HANDLING;
```

Comment:
These functions should not use special handling - this seems like a work-around for another work-around in `src/function/function_binder.cpp` where we don't return the output type but convert it to `SQLNULL` as a fallback.

Can we apply this diff to fix that, and turn this back into `DEFAULT_NULL_HANDLING`

```diff
diff --git a/src/function/function_binder.cpp b/src/function/function_binder.cpp
index 267f589d8f..87eacee685 100644
--- a/src/function/function_binder.cpp
+++ b/src/function/function_binder.cpp
@@ -306,13 +306,20 @@ unique_ptr<Expression> FunctionBinder::BindScalarFunction(ScalarFunctionCatalogE
        LogicalType return_type_if_null;
        switch (bound_function.return_type.id()) {
        case LogicalTypeId::ANY:
+               return_type_if_null = LogicalType::SQLNULL;
+               break;
        case LogicalTypeId::DECIMAL:
        case LogicalTypeId::STRUCT:
        case LogicalTypeId::LIST:
        case LogicalTypeId::MAP:
        case LogicalTypeId::UNION:
        case LogicalTypeId::ARRAY:
-               return_type_if_null = LogicalType::SQLNULL;
+               if (bound_function.return_type.AuxInfo()) {
+                       // this is a complete and valid type - return it
+                       return_type_if_null = bound_function.return_type;
+               } else {
+                       return_type_if_null = LogicalType::SQLNULL;
+               }
                break;
        default:
                return_type_if_null = bound_function.return_type;

```

Reply:
```
Wrong result in query! (test/sql/types/nested/map/test_null_map_interaction.test:27)!
================================================================================
SELECT TYPEOF(MAP_ENTRIES(NULL::MAP(TEXT, BIGINT)));
================================================================================
Mismatch on row 1, column 1
"NULL" <> STRUCT("key" VARCHAR, "value" BIGINT)[]
================================================================================
Expected result:
================================================================================
STRUCT("key" VARCHAR, "value" BIGINT)[]
================================================================================
Actual result:
================================================================================
"NULL"
```

Looks like that's not enough, I'll look into it

Reply:
This kicks in before we get to the `bind` function of the ScalarFunction
It happens after figuring out which overload to use, but before we run Bind

Reply:
I see - I suppose the nice solution here would be to wait for @Maxxen's rework on how type binding is handled as that would perhaps solve this, but we can merge this before that lands

Reply:
@Mytherin so should we merge this as-is then?
I think all the comments were related to DEFAULT_NULL_HANDLING

Reply:
Yes that's fine